From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Shrek Shao <shrekshao@google.com>
Date: Wed, 23 Sep 2026 13:36:35 -0700
Subject: [M152] [gl] Use blit for unsupported glReadPixels formats pt 1

Original change's description:
> [gl] Use blit for unsupported glReadPixels formats pt 1
>
> Under OpenGL ES, glReadPixels only guarantees a limited
> set of format/type combinations. Mostly format = GL_RGBA.
> Reading formats like R8Unorm, RG8Unorm with native format/type pair
> with glReadPixels is not guaranteed to succeed.
>
> This CL add toggles to emulate these copies with compute blits, for
> formats with existing blit implementation:
> - UseBlitForNonRGBAUnormTextureToBufferCopy
> - UseBlitForNonRGBAFloatTextureToBufferCopy
>
> Enable these toggles by default on the OpenGLES backend.
>
> Bug: 556959073
> Change-Id: Idc24c5736bbacc9176d34920946055b9ba332625
> Reviewed-on: https://dawn-review.googlesource.com/c/dawn/+/340475
> Commit-Queue: Shrek Shao <shrekshao@google.com>
> Reviewed-by: Geoff Lang <geofflang@chromium.org>
> Reviewed-by: Corentin Wallez <cwallez@chromium.org>

(cherry picked from commit 47a5943a2006839d0af8016074efa45db51ff334)

Bug: 564786090,556959073
Change-Id: Idc24c5736bbacc9176d34920946055b9ba332625

diff --git a/src/dawn/native/Buffer.cpp b/src/dawn/native/Buffer.cpp
index c31eac91d4fee7b0814701db54dc0758ed98d0e6..6ec71b0fb5da09ae2dbd7e396823d2c51c1d780a 100644
--- a/src/dawn/native/Buffer.cpp
+++ b/src/dawn/native/Buffer.cpp
@@ -188,6 +188,8 @@ wgpu::BufferUsage ComputeInternalBufferUsages(const DeviceBase* device,
             device->IsToggleEnabled(Toggle::UseBlitForStencilTextureToBufferCopy) ||
             device->IsToggleEnabled(Toggle::UseBlitForSnormTextureToBufferCopy) ||
             device->IsToggleEnabled(Toggle::UseBlitForBGRA8UnormTextureToBufferCopy) ||
+            device->IsToggleEnabled(Toggle::UseBlitForNonRGBAUnormTextureToBufferCopy) ||
+            device->IsToggleEnabled(Toggle::UseBlitForNonRGBAFloatTextureToBufferCopy) ||
             device->IsToggleEnabled(Toggle::UseBlitForRGB9E5UfloatTextureCopy) ||
             device->IsToggleEnabled(Toggle::UseBlitForRG11B10UfloatTextureCopy) ||
             device->IsToggleEnabled(Toggle::UseBlitForFloat16TextureCopy) ||
diff --git a/src/dawn/native/CommandEncoder.cpp b/src/dawn/native/CommandEncoder.cpp
index 003cf793b2ac23e6ab3612644feb0550a31c17c2..ffd64d1a047ce9b5595d17793f25b8998744f9e1 100644
--- a/src/dawn/native/CommandEncoder.cpp
+++ b/src/dawn/native/CommandEncoder.cpp
@@ -1159,6 +1159,19 @@ bool ShouldUseTextureToBufferBlit(const DeviceBase* device,
         device->IsToggleEnabled(Toggle::UseBlitForBGRA8UnormTextureToBufferCopy)) {
         return true;
     }
+    // Non-RGBA unorm
+    // TODO(crbug.com/561669028): *16Unorm also need to use blit for T2B copy on OpenGL
+    if ((format.format == wgpu::TextureFormat::R8Unorm ||
+         format.format == wgpu::TextureFormat::RG8Unorm) &&
+        device->IsToggleEnabled(Toggle::UseBlitForNonRGBAUnormTextureToBufferCopy)) {
+        return true;
+    }
+    // Non-RGBA float
+    if ((format.format == wgpu::TextureFormat::R32Float ||
+         format.format == wgpu::TextureFormat::RG32Float) &&
+        device->IsToggleEnabled(Toggle::UseBlitForNonRGBAFloatTextureToBufferCopy)) {
+        return true;
+    }
     // RGB9E5Ufloat
     if (format.format == wgpu::TextureFormat::RGB9E5Ufloat &&
         device->IsToggleEnabled(Toggle::UseBlitForRGB9E5UfloatTextureCopy)) {
diff --git a/src/dawn/native/Texture.cpp b/src/dawn/native/Texture.cpp
index 23c6752fe4b45b4b13391d4c5033d470801433b5..2092544d9ccfcbb9484b7b86d3209a07f9738ab6 100644
--- a/src/dawn/native/Texture.cpp
+++ b/src/dawn/native/Texture.cpp
@@ -564,6 +564,18 @@ bool CopySrcNeedsInternalTextureBindingUsage(const DeviceBase* device, const For
         device->IsToggleEnabled(Toggle::UseBlitForBGRA8UnormTextureToBufferCopy)) {
         return true;
     }
+    // Non-RGBA unorm
+    if ((format.format == wgpu::TextureFormat::R8Unorm ||
+         format.format == wgpu::TextureFormat::RG8Unorm) &&
+        device->IsToggleEnabled(Toggle::UseBlitForNonRGBAUnormTextureToBufferCopy)) {
+        return true;
+    }
+    // Non-RGBA float
+    if ((format.format == wgpu::TextureFormat::R32Float ||
+         format.format == wgpu::TextureFormat::RG32Float) &&
+        device->IsToggleEnabled(Toggle::UseBlitForNonRGBAFloatTextureToBufferCopy)) {
+        return true;
+    }
     // RGB9E5Ufloat
     if (format.format == wgpu::TextureFormat::RGB9E5Ufloat &&
         device->IsToggleEnabled(Toggle::UseBlitForRGB9E5UfloatTextureCopy)) {
diff --git a/src/dawn/native/Toggles.cpp b/src/dawn/native/Toggles.cpp
index 7254f7743e289496524974ff4ed5b77175d9e0da..fa12fded1e9a5d4a965108121ec0674d9fbec6ff 100644
--- a/src/dawn/native/Toggles.cpp
+++ b/src/dawn/native/Toggles.cpp
@@ -499,6 +499,16 @@ static constexpr ToggleEnumAndInfoList kToggleNameAndInfoList = {{
       "Use a shader based blit instead of a copy command to copy a buffer to a texture with "
       "supported format.",
       "https://crbug.com/dawn/348653642", ToggleStage::Device}},
+    {Toggle::UseBlitForNonRGBAUnormTextureToBufferCopy,
+     {"use_blit_for_non_rgba_unorm_texture_to_buffer_copy",
+      "Use a blit instead of a copy command to copy non-RGBA unorm textures to a buffer. "
+      "Workaround for OpenGLES.",
+      "https://issues.chromium.org/issues/556959073", ToggleStage::Device}},
+    {Toggle::UseBlitForNonRGBAFloatTextureToBufferCopy,
+     {"use_blit_for_non_rgba_float_texture_to_buffer_copy",
+      "Use a blit instead of a copy command to copy non-RGBA float textures to a buffer. "
+      "Workaround for OpenGLES.",
+      "https://issues.chromium.org/issues/556959073", ToggleStage::Device}},
     {Toggle::VulkanSplitBufferTextureCopyForArrayLayers,
      {"vulkan_split_buffer_texture_copy_for_array_layers",
       "Split a multi-layer buffer-to-texture / texture-to-buffer copy into one vkCmdCopy*Image "
diff --git a/src/dawn/native/Toggles.h b/src/dawn/native/Toggles.h
index 65086b6f663429ae5e3458d7fc91e92791a8cd98..1f9a67cf0d7e0c2d0e9d3127fa16e02d1ed8b6ec 100644
--- a/src/dawn/native/Toggles.h
+++ b/src/dawn/native/Toggles.h
@@ -127,6 +127,8 @@ enum class Toggle {
     UseBlitForFloat32TextureCopy,
     UseBlitForT2B,
     UseBlitForB2T,
+    UseBlitForNonRGBAUnormTextureToBufferCopy,
+    UseBlitForNonRGBAFloatTextureToBufferCopy,
     VulkanSplitBufferTextureCopyForArrayLayers,
     GLUseArrayLengthFromUniform,
     D3D11DisableCPUUploadBuffers,
diff --git a/src/dawn/native/opengl/PhysicalDeviceGL.cpp b/src/dawn/native/opengl/PhysicalDeviceGL.cpp
index 1f66c9fa173be7e4a8cde9494b861b0e19a20f75..02ce9dfac820e35fdffb468c9c0ba6da75e9893a 100644
--- a/src/dawn/native/opengl/PhysicalDeviceGL.cpp
+++ b/src/dawn/native/opengl/PhysicalDeviceGL.cpp
@@ -448,11 +448,8 @@ void PhysicalDevice::SetupBackendDeviceToggles(dawn::platform::Platform* platfor
     bool supportsStencilWriteTexture =
         gl.GetVersion().IsDesktop() || gl.IsGLExtensionSupported("GL_OES_texture_stencil8");
 
-    DAWN_ASSERT(gl.GetVersion().IsDesktop() || gl.IsAtLeastGLES(3, 2) ||
-                gl.IsGLExtensionSupported("GL_EXT_color_buffer_float"));
-    bool isFloat32Renderable = true;
-    bool isFloat16Renderable = true;
-    bool isRG11B10UfloatRenderable = true;
+    DAWN_CHECK(gl.GetVersion().IsDesktop() || gl.IsAtLeastGLES(3, 2) ||
+               gl.IsGLExtensionSupported("GL_EXT_color_buffer_float"));
 
     // TODO(crbug.com/dawn/343): Investigate emulation.
     deviceToggles->Default(Toggle::DisableIndexedDrawBuffers, !supportsIndexedDrawBuffers);
@@ -479,18 +476,26 @@ void PhysicalDevice::SetupBackendDeviceToggles(dawn::platform::Platform* platfor
     // For OpenGL ES, use compute shader blit to emulate bgra8unorm texture to buffer copies.
     deviceToggles->Default(Toggle::UseBlitForBGRA8UnormTextureToBufferCopy, !supportsBGRARead);
 
-    // For OpenGL ES, use compute shader blit to emulate rgb9e5ufloat texture to buffer copies.
+    // For OpenGL ES, use compute shader blit to emulate rgb9e5ufloat texture to buffer copies if
+    // not color-renderable.
     deviceToggles->Default(Toggle::UseBlitForRGB9E5UfloatTextureCopy, gl.GetVersion().IsES());
 
+    // For OpenGL ES, use compute shader blit to emulate texture to buffer copies to work around
+    // glReadPixels not guaranteed support for certain format/type combinations.
+    if (gl.GetVersion().IsES()) {
+        deviceToggles->Default(Toggle::UseBlitForNonRGBAUnormTextureToBufferCopy, true);
+        deviceToggles->Default(Toggle::UseBlitForNonRGBAFloatTextureToBufferCopy, true);
+        deviceToggles->Default(Toggle::UseBlitForFloat16TextureCopy, true);
+    }
+
     // Use compute shader blit to emulate rg11b10ufloat texture to buffer copies if not color
     // renderable.
-    deviceToggles->Default(Toggle::UseBlitForRG11B10UfloatTextureCopy, !isRG11B10UfloatRenderable);
-
-    // Use compute shader blit to emulate float16 texture to buffer copies if not color renderable.
-    deviceToggles->Default(Toggle::UseBlitForFloat16TextureCopy, !isFloat16Renderable);
+    // TODO(crbug.com/556959073): turn on after fix.
+    deviceToggles->Default(Toggle::UseBlitForRG11B10UfloatTextureCopy, false);
 
     // Use compute shader blit to emulate float32 texture to buffer copies if not color renderable.
-    deviceToggles->Default(Toggle::UseBlitForFloat32TextureCopy, !isFloat32Renderable);
+    // Note: GL_EXT_color_buffer_float is required so default to false.
+    deviceToggles->Default(Toggle::UseBlitForFloat32TextureCopy, false);
 
     // Use a blit to emulate stencil-only buffer-to-texture copies.
     deviceToggles->Default(Toggle::UseBlitForBufferToStencilTextureCopy, true);
diff --git a/src/dawn/tests/end2end/BufferZeroInitTests.cpp b/src/dawn/tests/end2end/BufferZeroInitTests.cpp
index 9a26777fb1aa48c902ba4998a4da290c2b63e768..25db757b2a848359d7e9c5a5a9409462514b175f 100644
--- a/src/dawn/tests/end2end/BufferZeroInitTests.cpp
+++ b/src/dawn/tests/end2end/BufferZeroInitTests.cpp
@@ -147,8 +147,12 @@ class BufferZeroInitTest : public DawnTest {
         wgpu::CommandBuffer commandBuffer = encoder.Finish();
 
         // TODO(b/513631768): SetInitialized is now skipped for use_blit_for_t2b path.
+        // Blit is also used for the R32Float format when
+        // use_blit_for_non_rgba_float_texture_to_buffer_copy is enabled.
         uint32_t expectedLazyClearCount = spec.lazyClearCount;
-        if (expectedLazyClearCount == 0u && (HasToggleEnabled("use_blit_for_t2b"))) {
+        if (expectedLazyClearCount == 0u &&
+            (HasToggleEnabled("use_blit_for_t2b") ||
+             HasToggleEnabled("use_blit_for_non_rgba_float_texture_to_buffer_copy"))) {
             expectedLazyClearCount = 1u;
         }
 
diff --git a/src/dawn/tests/end2end/CopyTests.cpp b/src/dawn/tests/end2end/CopyTests.cpp
index 29966663b1fa3520cd77952da98f3e0c93fe745a..f0163f6367b99dd78b0a1525d9ecdf4940d6338e 100644
--- a/src/dawn/tests/end2end/CopyTests.cpp
+++ b/src/dawn/tests/end2end/CopyTests.cpp
@@ -193,6 +193,10 @@ class CopyTests {
             case wgpu::TextureFormat::RG16Float:
             case wgpu::TextureFormat::RGBA16Float:
                 return GetExpectedTextureData16Float(layout);
+            case wgpu::TextureFormat::R32Float:
+            case wgpu::TextureFormat::RG32Float:
+            case wgpu::TextureFormat::RGBA32Float:
+                return GetExpectedTextureData32Float(layout);
             case wgpu::TextureFormat::RGB9E5Ufloat:
                 return GetExpectedTextureDataRGB9E5Ufloat(layout);
             case wgpu::TextureFormat::RG11B10Ufloat:
@@ -257,6 +261,56 @@ class CopyTests {
         return textureData;
     }
 
+    // Special function to generate test data for *32Float to avoid NaNs and subnormals
+    // that can be canonicalized or flushed to zero by shaders on some GPUs (e.g. Mali).
+    static std::vector<uint8_t> GetExpectedTextureData32Float(
+        const utils::TextureDataCopyLayout& layout) {
+        // These are some known 4-byte float values that are normal floats (non-NaN, non-subnormal)
+        // and always preserve their exact byte representation across texture load and store.
+        constexpr auto goodBytes = std::to_array<uint8_t>({
+            0x11, 0x22, 0x33, 0x3F,  // ~0.7
+            0x55, 0x66, 0x77, 0x40,  // ~3.86
+            0x99, 0xAA, 0xBB, 0x41,  // ~23.45
+            0x12, 0x34, 0x56, 0x42,  // ~53.55
+            0x78, 0x9A, 0xBC, 0x43,  // ~377.2
+            0xDE, 0xF0, 0x12, 0x44,  // ~587.7
+            0x34, 0x56, 0x78, 0x3E,  // ~0.24
+            0x90, 0x12, 0x34, 0xBF,  // ~ -0.7
+            0x56, 0x78, 0x9A, 0xC0,  // ~ -4.8
+            0xAB, 0xCD, 0xEF, 0xC1,  // ~ -29.9
+            0x10, 0x20, 0x30, 0x40,  // ~2.75
+            0x40, 0x50, 0x60, 0x41,  // ~14.02
+            0x70, 0x80, 0x90, 0x42,  // ~72.25
+            0x21, 0x43, 0x65, 0x43,  // ~229.26
+            0x87, 0xA9, 0xCB, 0x44,  // ~1629.3
+            0xED, 0x0F, 0x2E, 0x3E,  // ~0.17
+        });
+        constexpr uint32_t formatByteSize = 4;
+        constexpr uint32_t numGoodValues = goodBytes.size() / formatByteSize;
+        uint32_t bytesPerTexelBlock = layout.bytesPerRow / layout.texelBlocksPerRow;
+        uint32_t channelsPerTexel = bytesPerTexelBlock / formatByteSize;
+        std::vector<uint8_t> textureData(layout.byteLength);
+        for (uint32_t layer = 0; layer < layout.mipSize.depthOrArrayLayers; ++layer) {
+            const uint32_t byteOffsetPerSlice = layout.bytesPerImage * layer;
+            for (uint32_t y = 0; y < layout.mipSize.height; ++y) {
+                for (uint32_t x = 0; x < layout.mipSize.width; ++x) {
+                    for (uint32_t c = 0; c < channelsPerTexel; ++c) {
+                        uint32_t o = byteOffsetPerSlice + y * layout.bytesPerRow +
+                                     x * bytesPerTexelBlock + c * formatByteSize;
+                        uint32_t idx =
+                            formatByteSize *
+                            ((x * channelsPerTexel + c + 1 + (layer + 1) * y) % numGoodValues);
+                        textureData[o + 0] = goodBytes[idx + 0];
+                        textureData[o + 1] = goodBytes[idx + 1];
+                        textureData[o + 2] = goodBytes[idx + 2];
+                        textureData[o + 3] = goodBytes[idx + 3];
+                    }
+                }
+            }
+        }
+        return textureData;
+    }
+
     // Special function to generate test data for RGB9E5Ufloat to workaround nonunique encoding
     // issue.
     static std::vector<uint8_t> GetExpectedTextureDataRGB9E5Ufloat(
@@ -474,9 +528,12 @@ class CopyTests_T2B : public CopyTests_WithFormatParam {
 
         auto format = GetParam().mTextureFormat;
 
-        // TODO(crbug.com/dawn/2294): diagnose BGRA T2B failures on Pixel 4 OpenGLES
-        DAWN_SUPPRESS_TEST_IF(format == wgpu::TextureFormat::BGRA8Unorm && IsOpenGLES() &&
-                              IsAndroid() && IsQualcomm());
+        // TODO(crbug.com/dawn/2294): diagnose T2B failures on Pixel 4 OpenGLES
+        DAWN_SUPPRESS_TEST_IF((format == wgpu::TextureFormat::BGRA8Unorm ||
+                               format == wgpu::TextureFormat::R16Float ||
+                               format == wgpu::TextureFormat::RG16Float ||
+                               format == wgpu::TextureFormat::RGBA16Float) &&
+                              IsOpenGLES() && IsAndroid() && IsQualcomm());
 
         // TODO(dawn:1913): Many float formats tests failing for Metal backend on Mac Intel.
         DAWN_SUPPRESS_TEST_IF((format == wgpu::TextureFormat::R32Float ||
@@ -2146,57 +2203,65 @@ TEST_P(CopyTests_T2B, Texture3DMipUnaligned) {
     }
 }
 
-DAWN_INSTANTIATE_TEST_P(CopyTests_T2B,
-                        {D3D11Backend(), D3D11Backend({"d3d11_disable_map_on_default_buffers"}),
-                         D3D12Backend(), MetalBackend(), OpenGLBackend(), OpenGLESBackend(),
-                         OpenGLESBackend({"gl_defer"}), VulkanBackend(),
-                         VulkanBackend({"use_blit_for_snorm_texture_to_buffer_copy",
-                                        "use_blit_for_bgra8unorm_texture_to_buffer_copy"}),
-                         WebGPUBackend()},
-                        {
-                            wgpu::TextureFormat::R8Unorm,
-                            wgpu::TextureFormat::RG8Unorm,
-                            wgpu::TextureFormat::RGBA8Unorm,
+DAWN_INSTANTIATE_TEST_P(
+    CopyTests_T2B,
+    {D3D11Backend(), D3D11Backend({"d3d11_disable_map_on_default_buffers"}), D3D12Backend(),
+     MetalBackend(), OpenGLBackend(), OpenGLESBackend(), OpenGLESBackend({"gl_defer"}),
+     VulkanBackend(),
+     VulkanBackend({"use_blit_for_snorm_texture_to_buffer_copy",
+                    "use_blit_for_bgra8unorm_texture_to_buffer_copy"}),
+     WebGPUBackend()},
+    {
+        // Note: The formats below provide general T2B copy coverage across backends, and
+        // also record coverage for OpenGL compat toggles:
 
-                            wgpu::TextureFormat::R8Uint,
-                            wgpu::TextureFormat::R8Sint,
+        // Also covers OpenGL compat Toggle::UseBlitForNonRGBAUnormTextureToBufferCopy
+        wgpu::TextureFormat::R8Unorm,
+        wgpu::TextureFormat::RG8Unorm,
 
-                            wgpu::TextureFormat::R16Uint,
-                            wgpu::TextureFormat::R16Sint,
-                            wgpu::TextureFormat::R16Float,
+        wgpu::TextureFormat::RGBA8Unorm,
 
-                            wgpu::TextureFormat::RG16Uint,
-                            wgpu::TextureFormat::RG16Sint,
-                            wgpu::TextureFormat::RG16Float,
+        wgpu::TextureFormat::R8Uint,
+        wgpu::TextureFormat::R8Sint,
 
-                            wgpu::TextureFormat::R32Uint,
-                            wgpu::TextureFormat::R32Sint,
-                            wgpu::TextureFormat::R32Float,
+        wgpu::TextureFormat::R16Uint,
+        wgpu::TextureFormat::R16Sint,
+        wgpu::TextureFormat::R16Float,
 
-                            wgpu::TextureFormat::RG32Float,
-                            wgpu::TextureFormat::RG32Uint,
-                            wgpu::TextureFormat::RG32Sint,
+        wgpu::TextureFormat::RG16Uint,
+        wgpu::TextureFormat::RG16Sint,
+        wgpu::TextureFormat::RG16Float,
 
-                            wgpu::TextureFormat::RGBA16Uint,
-                            wgpu::TextureFormat::RGBA16Sint,
-                            wgpu::TextureFormat::RGBA16Float,
+        wgpu::TextureFormat::R32Uint,
+        wgpu::TextureFormat::R32Sint,
 
-                            wgpu::TextureFormat::RGBA32Float,
+        // Also covers OpenGL compat Toggle::UseBlitForNonRGBAFloatTextureToBufferCopy
+        wgpu::TextureFormat::R32Float,
+        wgpu::TextureFormat::RG32Float,
 
-                            wgpu::TextureFormat::RGB10A2Unorm,
-                            wgpu::TextureFormat::RG11B10Ufloat,
+        wgpu::TextureFormat::RG32Uint,
+        wgpu::TextureFormat::RG32Sint,
 
-                            // Testing OpenGL compat Toggle::UseBlitForRGB9E5UfloatTextureCopy
-                            wgpu::TextureFormat::RGB9E5Ufloat,
+        wgpu::TextureFormat::RGBA16Uint,
+        wgpu::TextureFormat::RGBA16Sint,
+        wgpu::TextureFormat::RGBA16Float,
 
-                            // Testing OpenGL compat Toggle::UseBlitForSnormTextureToBufferCopy
-                            wgpu::TextureFormat::R8Snorm,
-                            wgpu::TextureFormat::RG8Snorm,
-                            wgpu::TextureFormat::RGBA8Snorm,
+        wgpu::TextureFormat::RGBA32Float,
 
-                            // Testing OpenGL compat Toggle::UseBlitForBGRA8UnormTextureToBufferCopy
-                            wgpu::TextureFormat::BGRA8Unorm,
-                        });
+        wgpu::TextureFormat::RGB10A2Unorm,
+        wgpu::TextureFormat::RG11B10Ufloat,
+
+        // Also covers OpenGL compat Toggle::UseBlitForRGB9E5UfloatTextureCopy
+        wgpu::TextureFormat::RGB9E5Ufloat,
+
+        // Also covers OpenGL compat Toggle::UseBlitForSnormTextureToBufferCopy
+        wgpu::TextureFormat::R8Snorm,
+        wgpu::TextureFormat::RG8Snorm,
+        wgpu::TextureFormat::RGBA8Snorm,
+
+        // Also covers OpenGL compat Toggle::UseBlitForBGRA8UnormTextureToBufferCopy
+        wgpu::TextureFormat::BGRA8Unorm,
+    });
 
 class CopyTests_T2B_No_Format_Param : public CopyTests, public DawnTest {};
 
@@ -3313,6 +3378,8 @@ TEST_P(CopyTests_T2T, CopyFromNonZeroMipLevelWithTexelBlockSizeLessThan4Bytes) {
     DAWN_SUPPRESS_TEST_IF(IsImgTec());
     // TODO(crbug.com/473582006): [Capture] issue calling ResolveDeferredExpectationsNow.
     DAWN_SUPPRESS_TEST_IF(IsCaptureReplayCheckingEnabled());
+    // TODO(crbug.com/dawn/2294): diagnose T2B failures on Pixel 4 OpenGLES
+    DAWN_SUPPRESS_TEST_IF(IsOpenGLES() && IsAndroid() && IsQualcomm());
 
     constexpr std::array<wgpu::TextureFormat, 11> kFormats = {
         {wgpu::TextureFormat::RG8Sint, wgpu::TextureFormat::RG8Uint, wgpu::TextureFormat::RG8Snorm,
diff --git a/src/dawn/tests/end2end/NonzeroTextureCreationTests.cpp b/src/dawn/tests/end2end/NonzeroTextureCreationTests.cpp
index bd214b76b52321916c8df4c9f97459ab5370ff78..24654eded33dbb3602cc9a68e5499f38cd3cfa35 100644
--- a/src/dawn/tests/end2end/NonzeroTextureCreationTests.cpp
+++ b/src/dawn/tests/end2end/NonzeroTextureCreationTests.cpp
@@ -120,6 +120,12 @@ class NonzeroTextureCreationTests : public DawnTestWithParams<Params> {
         // TODO(dawn:1844): Work around this by clearing layers one by one on Intel.
         DAWN_SUPPRESS_TEST_IF(IsD3D11() && IsIntel() && GetParam().mDepthOrArrayLayers > 6);
 
+        // TODO(crbug.com/556959073): BlitTextureToBuffer cube sampling fails on Pixel 10 OpenGLES
+        // for RG8Unorm small mips.
+        DAWN_SUPPRESS_TEST_IF(IsImgTec() && IsOpenGLES() &&
+                              GetParam().mFormat == wgpu::TextureFormat::RG8Unorm &&
+                              GetParam().mDepthOrArrayLayers == 6 && GetParam().mMip >= 2);
+
         wgpu::TextureDescriptor descriptor;
         descriptor.dimension = GetParam().mDimension;
         descriptor.size.width = kSize;
